### PR TITLE
Feat/4/identity server integration

### PR DESCRIPTION
**Pull request by Zakerman110**
Added oidc-client for angular to work with OpenId and Oauth, secured Units.API with Bearer token, added interceptor to apply tokens for each Http request

### CODE REVIEW

File: MilitaryManager.IdentityServer/Startup.cs [hunk 1]
```diff
@@ -84,10 +86,11 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
...
                 builder => builder
-                    .WithOrigins(
-                        "http://localhost:4200",
-                        "http://localhost:5001",
-                        "http://localhost:5000")
+                    //.WithOrigins(
+                    //    "http://localhost:4200",
+                    //    "https://localhost:5001",
+                    //    "http://localhost:5000")
+                    .AllowAnyOrigin()
```
Comment (YevheniiVovk) [inline MilitaryManager.IdentityServer/Startup.cs]: Why we change to allow to all?

File: MilitaryManager.Units.API/Controllers/WeatherForecastController.cs [hunk 2]
```diff
@@ -24,16 +25,20 @@ public WeatherForecastController(ILogger<WeatherForecastController> logger)
         }
 
         [HttpGet]
-        public IEnumerable<WeatherForecast> Get()
+        [Authorize]
+        //public IEnumerable<WeatherForecast> Get()
```
Comment (YevheniiVovk) [inline MilitaryManager.Units.API/Controllers/WeatherForecastController.cs]: What a reason of modify current controller? 

Please do not leave any commented code

File: MilitaryManager.Units.API/Controllers/WeatherForecastController.cs [hunk 3]
```diff
@@ -24,16 +25,20 @@ public WeatherForecastController(ILogger<WeatherForecastController> logger)
...
             var rng = new Random();
             return Enumerable.Range(1, 5).Select(index => new WeatherForecast
             {
+                id = index,
                 Date = DateTime.Now.AddDays(index),
                 TemperatureC = rng.Next(-20, 55),
                 Summary = Summaries[rng.Next(Summaries.Length)]
             })
-            .ToArray();
+            //.ToArray();
```
Comment (YevheniiVovk) [inline MilitaryManager.Units.API/Controllers/WeatherForecastController.cs]: Please do not leave any commented code

File: MilitaryManager.Units.API/Controllers/WeatherForecastController.cs [hunk 4]
```diff
@@ -24,16 +25,20 @@ public WeatherForecastController(ILogger<WeatherForecastController> logger)
...
+            .ToArray().FirstOrDefault();
```
Comment (YevheniiVovk) [inline MilitaryManager.Units.API/Controllers/WeatherForecastController.cs]: Why does get with NO parameters return only one object

File: MilitaryManager.Units.API/Startup.cs [hunk 5]
```diff
@@ -19,6 +26,37 @@ public Startup(IConfiguration configuration)
...
             services.AddControllers();
+
+            services.AddAuthentication(JwtBearerDefaults.AuthenticationScheme)
+           .AddJwtBearer(opt =>
+           { 
+               opt.RequireHttpsMetadata = false;
+               
+               //opt.Authority = $"https://{{Configuration.GetConnectionString('DefaultConnection')}}:5007";
+               //TODO: move link to appsettings
+               opt.Authority = "http://militarymanager.identity.server";
```
Comment (YevheniiVovk) [inline MilitaryManager.Units.API/Startup.cs]: What is this url how it should works ?

File: MilitaryManager.Units.API/Startup.cs [hunk 6]
```diff
@@ -19,6 +26,37 @@ public Startup(IConfiguration configuration)
...
+               //opt.Authority = "http://localhost:5006";
+               opt.TokenValidationParameters = new TokenValidationParameters
+               {
+                   //TODO: IT IS NOT SECURE
+                   ValidateAudience = false,
+                   ValidateIssuer = false,
+               };
```
Comment (YevheniiVovk) [inline MilitaryManager.Units.API/Startup.cs]: IT IS NOT SECURE

File: MilitaryManager.Units.API/Startup.cs [hunk 7]
```diff
@@ -19,6 +26,37 @@ public Startup(IConfiguration configuration)
...
+               //opt.TokenValidationParameters = new TokenValidationParameters
+               //{
+               //    ValidateIssuerSigningKey = true,
+               //    ValidateLifetime = true,
+               //    ValidateIssuer = true,
+               //    ValidIssuer = "api1",  
+               //    ValidateAudience = true
+               //};
+               //opt.Audience = "api1";
+               //opt.Configuration = new OpenIdConnectConfiguration();
```
Comment (YevheniiVovk) [inline MilitaryManager.Units.API/Startup.cs]: Code should be uncommented and not needed code should be remove

File: MilitaryManager/Startup.cs [hunk 8]
```diff
@@ -71,7 +71,8 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
 
             app.UseCors(
                 builder => builder
-                    .WithOrigins("http://localhost:4200", "http://localhost:5001", "http://localhost:5000")
+                    //.WithOrigins("http://localhost:4200", "http://localhost:5001", "http://localhost:5000", "https://localhost:5005")
```
Comment (YevheniiVovk) [inline MilitaryManager/Startup.cs]: Why 5005 port is unavailable ?

File: MilitaryManager/client-app/src/app/shared/constants.ts [hunk 9]
```diff
@@ -0,0 +1,6 @@
+export class Constants {
+    public static apiRoot = "https://localhost:5001/api";
+    public static clientRoot = "https://localhost:5001";
+    public static idpAuthority = "https://localhost:5007"
+    public static clientId = "js";
+}
```
Comment (YevheniiVovk) [inline MilitaryManager/client-app/src/app/shared/constants.ts]: End of file

File: MilitaryManager/client-app/src/environments/environment.ts [hunk 10]
```diff
@@ -5,7 +5,8 @@
 export const environment = {
   production: false,
   webserverApiPath: '/api/',
-  webUrl: 'https://localhost:5001'
+  webUrl: 'https://localhost:5001',
+  unitsApiUrl: 'https://localhost:5005'
```
Comment (YevheniiVovk) [inline MilitaryManager/client-app/src/environments/environment.ts]: We do not need it here

File: docker-compose.override.yml [hunk 11]
```diff
@@ -37,7 +37,7 @@ services:
     environment:
       - ASPNETCORE_ENVIRONMENT=Development
       - ASPNETCORE_URLS=https://+:443;http://+:80
-      - ConnectionStrings__DefaultConnection=Server=host.docker.internal;Database=IdentityServer_AspNet_Identity;User ID=Gwa_System;Password=T&t137S5tP!+3SMJQ<r0KJ;MultipleActiveResultSets=true;Application Name=IdentityServer
+      - ConnectionStrings__DefaultConnection=Server=host.docker.internal;Database=IdentityServer_AspNet_Identity;User ID=javist;Password=123456java;MultipleActiveResultSets=true;Application Name=IdentityServer
```
Comment (YevheniiVovk) [inline docker-compose.override.yml]: Please do not commit your changes to docker compose

File: MilitaryManager/client-app/src/app/shared/services/auth-interceptor.service.ts [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+      this._authService.getAccessToken()
+      .then(token => {
+        // overrides any custom headers
+        const headers = new HttpHeaders().set('Authorization', `Bearer ${token}`);
+        const authRequest = req.clone({ headers });
+        return next.handle(authRequest).toPromise();
+      })
+    )
+  }
+}
```
Comment (YevheniiVovk) [inline MilitaryManager/client-app/src/app/shared/services/auth-interceptor.service.ts:24]: We try to make cookies auth but make an implementation of token, what a reason ?

File: MilitaryManager/client-app/src/app/shared/constants.ts [hunk 13]
```diff
@@ -0,0 +1,6 @@
+export class Constants {
+    public static apiRoot = "https://localhost:5001/api";
+    public static clientRoot = "https://localhost:5001";
+    public static idpAuthority = "https://localhost:5007"
+    public static clientId = "js";
```
Comment (YevheniiVovk) [inline MilitaryManager/client-app/src/app/shared/constants.ts]: We do not need it here

File: MilitaryManager/client-app/src/app/app.component.ts [hunk 14]
```diff
@@ -10,15 +11,34 @@ import { ClientConfigurationService } from './shared/services/core/client-config
...
     constructor(
         private configService: ConfigService,
         private clientConfigService: ClientConfigurationService,
-        private apiService: ApiService) {
+        private apiService: ApiService,
+        private _authService: AuthService) {
+        this._authService.loginChanged
+        .subscribe(userAuthenticated => {
+        this.userAuthenticated = userAuthenticated;
+        })
```
Comment (YevheniiVovk) [inline MilitaryManager/client-app/src/app/app.component.ts]: Formatting

File: MilitaryManager/client-app/src/app/app.component.ts [hunk 15]
```diff
@@ -10,15 +11,34 @@ import { ClientConfigurationService } from './shared/services/core/client-config
...
     }
 
     ngOnInit(): void {
         this.configService.single.get().subscribe((data: ConfigModel) => {
             this.clientConfigService.config = data;
         });
+        this._authService.isAuthenticated()
+        .then(userAuthenticated => {
+        this.userAuthenticated = userAuthenticated;
```
Comment (YevheniiVovk) [inline MilitaryManager/client-app/src/app/app.component.ts]: Formatting

File: MilitaryManager.Units.API/Startup.cs [hunk 16]
```diff
@@ -19,6 +26,37 @@ public Startup(IConfiguration configuration)
         public void ConfigureServices(IServiceCollection services)
         {
             services.AddControllers();
+
+            services.AddAuthentication(JwtBearerDefaults.AuthenticationScheme)
+           .AddJwtBearer(opt =>
```
Comment (YevheniiVovk) [inline MilitaryManager.Units.API/Startup.cs]: Formatting

File: MilitaryManager/client-app/src/app/components/signout-redirect-callback/signout-redirect-callback.component.ts [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+export class SignoutRedirectCallbackComponent implements OnInit {
+
+  constructor(private _authService: AuthService, private _router: Router) { }
+
+  ngOnInit(): void {
+    this._authService.finishLogout()
+    .then(_ => {
+      this._router.navigate(['/'], { replaceUrl: true });
+    })
+  }
```
Comment (YevheniiVovk) [inline MilitaryManager/client-app/src/app/components/signout-redirect-callback/signout-redirect-callback.component.ts]: Formatting

File: MilitaryManager/client-app/src/app/components/test1/test1.component.ts [hunk 18]
```diff
@@ -7,9 +9,15 @@ import { Component, OnInit } from '@angular/core';
...
-  constructor() { }
+  constructor(public apiService: ApiService) { }
 
   ngOnInit(): void {
   }
 
+  getWeatherUnits() {
+    this.apiService.units.single.get().subscribe(res => {
+      console.log(res);
+    })
```
Comment (YevheniiVovk) [inline MilitaryManager/client-app/src/app/components/test1/test1.component.ts]: Console log it is not good idea to live it in code

File: MilitaryManager/client-app/src/app/shared/services/auth.service.ts [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+  public isAuthenticated = (): Promise<boolean> => {
+    return this._userManager.getUser()
+    .then(user => {
+      if(this._user !== user){
+        // @ts-ignore
+        this._loginChangedSubject.next(this.checkUser(user));
+      }
+      // @ts-ignore
+      this._user = user;
+      // @ts-ignore
```
Comment (YevheniiVovk) [inline MilitaryManager/client-app/src/app/shared/services/auth.service.ts]: Why do we need @ts-ignore

File: MilitaryManager.IdentityServer/IdentityConfig.cs [hunk 20]
```diff
@@ -58,10 +59,19 @@ public static IEnumerable<ApiResource> GetApiResources()
...
 
 		}
+
+		public static IEnumerable<ApiScope> GetApiScopes()
+        {
+			return new List<ApiScope> 
+			{ 
+				new ApiScope("api1", "API 1") 
+			};
+		}
```
Comment (YevheniiVovk) [inline MilitaryManager.IdentityServer/IdentityConfig.cs]: Formatting

File: MilitaryManager/client-app/src/app/shared/services/auth-interceptor.service.ts (hunk 12 above)
Comment (Zakerman110) [reply MilitaryManager/client-app/src/app/shared/services/auth-interceptor.service.ts:24]: Cookies are not attaching to requests

File: MilitaryManager.Units.API/Startup.cs [hunk 21]
```diff
@@ -52,11 +56,74 @@ public void ConfigureServices(IServiceCollection services)
...
+                            string access_token = data["access_token"].Value<string>();
+                            context.Token = access_token;
+                            return Task.CompletedTask;
+                        }
+                    };
+                });
 
-            services.AddAuthorization();
+            // Configure cookie authentication.
+            /* services.AddAuthentication(CookieAuthenticationDefaults.AuthenticationScheme)
```
Comment (5haki6) [inline MilitaryManager.Units.API/Startup.cs]: is it good practice to leave too many commented code in pr?

**Review by 5haki6 — CHANGES_REQUESTED** [review]

File: MilitaryManager.Units.API/Startup.cs (hunk 21 above)
Comment (Zakerman110) [reply MilitaryManager.Units.API/Startup.cs]: These comments have already been removed, you are viewing an old commit - c39989f

**Review by 5haki6 — APPROVED** [review]

**Review by vladtymo — APPROVED** [review]

**Pull request merged**